20220831-add-type-restrictions-to-json-syntax.md

```diff
@@ -0,0 +1,469 @@
+
+## What is the expected outcome?
+
+- Using the extra type information we get from this change, we will be able to more easily traverse the graph in reverse, thus allowing us to provide a solution for reverse expansion and optimizing the ListObjects endpoint.
+- Users reading an authorization model will be able to better interpret its meaning (for example by understanding that repository owners must be either users or group members).
+
+# What it is?
+This RFC proposes an update to the OpenFGA JSON syntax requiring users to indicate all the types of users that could be **directly related** to an object of a certain type through a particular relation.
+
+For example,
+- parents of a document have to be objects of type folder
+- owners of a repository have to be of type user or a userset of group members
+
+We will add an `assignable_relations` to each relation to indicate what types of users can be directly related to it.
+
+`assignable_relations` will be an array of objects, each object must have a type and an optional relation.
+- having the following `"parent": { "assignable_relations": [{ "type": "group" }] }` in the group type definition indicates that only users of type group can be directly related to a group as parent
+- having the following `"member": { "assignable_relations": [{ "type": "user" }, { "type": "group", "relation": "member" }] }` in the group type definition indicates that only objects of type user or usersets of type group and relation member can be directly related to a group as member
+- having the following `"member": { "assignable_relations": [{ "type": "user:*" }, { "type": "employee:*" }] }` in the group type definition indicates that the `*` syntax is allowed and that when present, it means all objects of type user or group can be directly related to a group as member
+
+This will affect only relations that are [directly related](https://openfga.dev/docs/modeling/building-blocks/direct-relationships) (as in they are considered "assignable" and have [the direct relationship keyword ("this")](https://openfga.dev/docs/configuration-language#the-direct-relationship-keyword) in their relation definition).
+For relation definitions that:
+- have the direct relationship keyword (this): this key must be present in `assignable_relations` and MUST be an non-empty array containing at least one type/relation combination
+- do not have the direct relationship keyword: this key must be present in `assignable_relations` and MUST be an empty array
```

Comment:
Curious why the user needs to define an empty array when we know that there is no direct relationship

Reply:
Will make it optional